Add cloud provider integration doc for CCCMO (#67)
openshift/cluster-cloud-controller-manager-operator: docs/dev/cloud-provider-integration.md

The pod is required to use bootstrap kubeconfig, so you are required to specify a hostPath volume pointing to `/etc/kubernetes/bootstrap-secrets` and specify a `--kubeconfig `flag.

Your CCM might require a cloud-config to work. Bootstrap part of the operator will extract cloud-config from ConfigMap generated by [cluster-config-operator](https://github.com/openshift/cluster-config-operator) and place it into the  `/etc/kubernetes/bootstrap-configs/cloud.conf` file. You will be required to mount it into the pod, and pass the --cloud-config flag to it.

During development, the Pod image you specify can be any image, ie a public image that can be used when you `oc apply -f pod.yaml.` However, we can’t use this image in a production cluster, so we are required to substitute the image for a Red Hat managed image in customer clusters. In order for substitution to work you have to name your container as `cloud-controller-manager.`

```
TBD: Process for passing cloud credentials secret in bootstrap
```

> In the bootstrap part of the installation procedure we only require to run the `cloud-node` controller. `--controllers=cloud-node` would restrict your CCM to run only that controller. This controller is responsible for untainting Nodes from n`ode.cloudprovider.kubernetes.io/uninitialized` taint, and setting Node IP addresses, as well as labeling nodes with labels required by autoscaler, storage, etc., and setting the Node `spec.ProviderID.`

Reviewer: ```suggestion
In the bootstrap part of the installation procedure we only require to run the `cloud-node` controller. `--controllers=cloud-node` would restrict your CCM to run only that controller. This controller is responsible for untainting Nodes from `node.cloudprovider.kubernetes.io/uninitialized` taint, and setting Node IP addresses, as well as labelling nodes with labels required by scheduler, autoscaler, storage, etc., and setting the Node `spec.ProviderID.`
```

     - effect: NoSchedule
       key: node-role.kubernetes.io/master
       operator: Exists
     - effect: NoExecute
       key: node.kubernetes.io/unreachable
       operator: Exists
       tolerationSeconds: 120
     # This taint is strictly required in order for CCM to be scheduled
     # on the master Node at the moment when the cluster
     # lacks CCM replicas to untaint the Node
     - effect: NoSchedule
       key: node.cloudprovider.kubernetes.io/uninitialized
       value: "true"
```

Our operator is responsible for synchronization of `cloud-config` ConfigMap from `openshift-config` and `openshift-config-managed` namespace to the namespace where the CCM resources are provisioned.  The ConfigMap is named `cloud-conf `and could be mounted into a CCM pod for later use if your cloud provider requires it.

Credentials secret serving to `openshift-cloud-controller-manager` namespace is carried by [https://github.com/openshift/cloud-credential-operator](https://github.com/openshift/cloud-credential-operator) for us. You need to implement your cloud provider support there, and add a `CredentialsRequest` resource in `manifests` directory. 

> ## Cloud-provider fork on openshift side

Reviewer 2: ```suggestion
## Cloud-provider fork on OpenShift side
```

> You are required to create your cloud-provider fork under openshift organization. This fork will be responsible for building and resolving your provider images, as well as following openshift release branching cadence.That repository has to be added into CI system and will run post submit and periodic jobs with e2e tests on your cloud-provider.

Reviewer 2: ```suggestion
You are required to create your cloud-provider fork under OpenShift organization. This fork will be responsible for building and resolving your provider images, as well as following OpenShift release branching cadence.That repository has to be added into CI system and will run post submit and periodic jobs with e2e tests on your cloud-provider.
```

You will require repository addition to [openshift/release](https://github.com/openshift/release). Here is an [PR](https://github.com/openshift/release/pull/17199) for AWS [openshift/cloud-provider-aws](https://github.com/openshift/cloud-provider-aws) repository, establishing initial set of CI jobs, including `images` job responsible for building and tagging `registry.ci.openshift.org/openshift:aws-cloud-controller-manager`

Once an image is added to the CI system, it will be present in CI builds but not in release builds. You have to follow ART guidelines and create tickets to add your cloud-provider repository into release. See [https://docs.ci.openshift.org/docs/how-tos/onboarding-a-new-component/#product-builds-and-becoming-part-of-an-openshift-release-image](https://docs.ci.openshift.org/docs/how-tos/onboarding-a-new-component/#product-builds-and-becoming-part-of-an-openshift-release-image) You need openshift engineering help to set this up.

> It is very important to get this done step before the next one,an addition to image-references could be done only once the ART request has been resolved, or it will break nightly releases. We currently track relevant requests in this [doc](https://docs.google.com/document/d/1XB_-I9wTBwsGw-a7rwiCmOlnxTVnUdbw5odkyLKwp1k/edit?usp=sharing).

Reviewer 2: ```suggestion
It is very important to get this step done before the next one,an addition to image-references could be done only once the ART request has been resolved, or it will break nightly releases. We currently track relevant requests in this [doc](https://docs.google.com/document/d/1XB_-I9wTBwsGw-a7rwiCmOlnxTVnUdbw5odkyLKwp1k/edit?usp=sharing).
```

```yaml
- name: aws-cloud-controller-manager
  from:
   kind: DockerImage
   name: registry.ci.openshift.org/openshift:aws-cloud-controller-manager
```

You will have to extend the config [images](https://github.com/openshift/cluster-cloud-controller-manager-operator/blob/master/manifests/0000_26_cloud-controller-manager-operator_01_images.configmap.yaml) list with image name:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
 name: cloud-controller-manager-images
 namespace: openshift-cloud-controller-manager-operator
data:
 images.json: >
   {
     "cloudControllerManagerAWS": "registry.ci.openshift.org/openshift:aws-cloud-controller-manager",
   … # Other images 
   }
```

After that you need to ensure the `imageReferences`  [contains](https://github.com/openshift/cluster-cloud-controller-manager-operator/blob/c161640ef47e232df5c9a4c9298ba95551fa48d9/pkg/config/config.go#L18-L23) your image, and is later used in substitution. 

## Manifests representation

We use a couple of substitutions in manifests you place for the files in the `assets` or `bootstrap` folders.

> 

Reviewer 2: blank lines
